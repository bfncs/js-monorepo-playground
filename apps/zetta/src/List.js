import React from 'react';
import PropTypes from 'prop-types';
import Item from './Item';

const List = ({ items, setItemDone }) => (
  <ul className={'list'}>
    {items.map(({ id, title, done }) => (
      <li key={id}>
        <Item title={title} done={done} setDone={nextDone => setItemDone(id, nextDone)} />
      </li>
    ))}
  </ul>
);

List.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired,
  })).isRequired,
  setItemDone: PropTypes.func.isRequired,
};

export default List;