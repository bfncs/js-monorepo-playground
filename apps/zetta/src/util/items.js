import shortId  from 'shortid';

export const createItem = (title, done = false) => ({
  id: shortId.generate(),
  title,
  done,
});

const getAndUpdateItem = (items = [], id, fn) =>
  items.map(item => item.id === id ? fn(item) : item);

export const setItemDone = (items = [], id, done) =>
  getAndUpdateItem(items, id, item => ({ ...item, done }));