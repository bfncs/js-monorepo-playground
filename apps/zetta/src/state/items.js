import { withItemDone } from '../util/items';

const SET_STATE_ITEM_DONE = 'SET_STATE_ITEM_DONE';
const ADD_NEW_ITEM = 'ADD_NEW_ITEM';

export const setStateItemDone = (id, done) => ({
  type: SET_STATE_ITEM_DONE,
  id,
  done,
});

export const addNewItem = ({ id, title, done = false }) =>({
  type: ADD_NEW_ITEM,
  id,
  title,
  done,
});

export const reducer = (items = [], action) => {
  switch (action.type) {
    case SET_STATE_ITEM_DONE:
      return withItemDone(items, action.id, action.done);
    case ADD_NEW_ITEM:
      const { id, title, done } = action;
      return [...items, { id, title, done }];
    default:
      return items;
  }
};