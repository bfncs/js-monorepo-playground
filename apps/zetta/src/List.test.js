import React from 'react';
import renderer from 'react-test-renderer';
import List from './List';
import { createItem } from './util/items';

const NO_OP = () => {};
const ITEMS = ['foo', 'bar', 'baz'].map((title, count) => createItem(title, count % 2 === 0));

describe('List snapshot tests', () => {

  it('matches snapshot when empty', () => {
    expect(renderer.create(
      <List items={[]} setItemDone={NO_OP}/>
    ).toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with items', () => {
    expect(renderer.create(
      <List items={ITEMS} setItemDone={NO_OP}/>
    ).toJSON()).toMatchSnapshot();
  });

});