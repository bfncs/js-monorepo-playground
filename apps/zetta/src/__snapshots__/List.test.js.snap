// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List snapshot tests matches snapshot when empty 1`] = `
<ul
  className="list"
/>
`;

exports[`List snapshot tests matches snapshot with items 1`] = `
<ul
  className="list"
>
  <li>
    <div>
      <input
        checked={true}
        onClick={[Function]}
        type="checkbox"
      />
      <span>
        foo
      </span>
    </div>
  </li>
  <li>
    <div>
      <input
        checked={false}
        onClick={[Function]}
        type="checkbox"
      />
      <span>
        bar
      </span>
    </div>
  </li>
  <li>
    <div>
      <input
        checked={true}
        onClick={[Function]}
        type="checkbox"
      />
      <span>
        baz
      </span>
    </div>
  </li>
</ul>
`;
